Generated blog post summary for EP248 (#2653)
quarkusio/quarkusio.github.io: _posts/2026-05-19-quarkus-insights-248-ddd-hexagonal-architecture.adoc

---
layout: post
title: 'Quarkus Insights #248: Introduction to Domain-Driven Design and Hexagonal Architecture'
date: 2026-05-19
tags: insights ddd architecture design-patterns quarkus
synopsis: 'Jeremy Davis introduces Domain-Driven Design and Hexagonal Architecture patterns with Quarkus, demonstrating how to structure applications for maintainability and exploring AI-assisted development workflows.'
author: jcobb
---

This summary was generated using AI, reviewed by humans - https://youtube.com/live/2PnZ1zzuWn0[watch the video] for the full story.

= Quarkus Insights #248: Introduction to Domain-Driven Design and Hexagonal Architecture

In episode 248 of Quarkus Insights, https://www.linkedin.com/in/jeremyrdavis/[Jeremy Davis], a software engineer at Red Hat with extensive experience in the JBoss ecosystem, provided an in-depth introduction to https://martinfowler.com/bliki/DomainDrivenDesign.html[Domain-Driven Design (DDD)] and Hexagonal Architecture (also known as Ports and Adapters) using https://quarkus.io/[Quarkus].

> 

Reviewer: Jeremy isn't a software engineer. I would also drop "at Red Hat".

Author: Changed to "Solutions Architect" and removed Red Hat reference.

> 

Reviewer: I'd drop `(also known as Ports and Adapters)`

Author: Dropped.

++++
<iframe width="560" height="315" src="https://www.youtube.com/embed/2PnZ1zzuWn0?si=VbQcY8LPfr6Yf-cX" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen></iframe>
++++

== Why DDD is Having a Moment

Jeremy opened by noting that DDD seems to be experiencing renewed interest, with multiple presentation requests recently. He theorizes this resurgence relates to the rise of AI-assisted development and agentic systems, where DDD's structured approach to organizing business logic makes it easier for AI tools to understand and generate code.

== The Core Principles of Domain-Driven Design

=== Ubiquitous Language

One of Jeremy's favorite aspects of DDD is the emphasis on **ubiquitous language** - a shared vocabulary between developers and business stakeholders. He illustrated this with the restaurant industry example: "86 ketchup" means "we're out of ketchup" to anyone who's worked in restaurants, but is completely nonsensical to outsiders.

> The key insight: every industry has its own language, and DDD encourages teams to:

Reviewer 2: ```suggestion
Since every industry has its own language, DDD encourages teams to:
```

The original phrasing was quite slop-py (which is fair enough, since it's AI-generated). But I think a little adjustment makes it more human.

- Identify and document domain-specific terminology
- Use that language consistently in code, conversations, and documentation
- Refine the language iteratively with each sprint or development cycle

=== Domains and Subdomains

> For the demo, Jeremy built a Quarkus Insights episode scheduling application with several subdomains:

Reviewer 2: ```suggestion
For the demo, Jeremy built an application for scheduling Quarkus Insights episodes, with several subdomains:
```

I had to read the original phrasing twice to parse it properly.

Each subdomain represents a distinct area of business concern with its own models and logic.

== Building Blocks of DDD

=== Aggregates

**Aggregates** are the core business concepts that encapsulate business rules (invariants). In Jeremy's example, the `EpisodeAggregate` serves as the root entity:

[source,java]
----
public class EpisodeAggregate {
    private EpisodeId id;
    private EpisodeTitle title;
    private String description;
    private EpisodeAirDate airDate;
    private Collection<DomainEvent> domainEvents;
    
    public static EpisodeAggregate schedule(
        EpisodeTitle title, 
        String description, 
        EpisodeAirDate airDate) {
        // Business logic here
        EpisodeAggregate episode = new EpisodeAggregate();
        episode.domainEvents.add(new EpisodeScheduledEvent());
        return episode;
    }
}
----

Key characteristics:

- Aggregates are **POJOs** with no framework dependencies
- All business logic lives in the aggregate
- Aggregates maintain **invariants** (rules that must always be true)
- You interact with the object graph through the aggregate root

=== Value Objects

> **Value objects** encapsulate data with no identity. Examples include addresses, email addresses, or in Jeremy's demo, episode titles:

Reviewer 2: ```suggestion
**Value objects** encapsulate data with no identity. That is,  a value object is defined entirely by its attributes, not by some unique ID. Examples include addresses, email addresses, or in Jeremy's demo, episode titles:
```

Original was not wrong, but I didn't follow it, so I expanded :)

[source,java]
----
public record EpisodeTitle(String value) {
    public EpisodeTitle {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException("Episode title cannot be null or blank");
        }
        // Could add more validation:
        // - Must contain "Quarkus"
        // - Must contain "Insights"
        // - Must have episode number
    }
}
----

Value objects enable validation at creation time and make the domain model more expressive.

=== Domain Events

**Domain events** represent facts that the business cares about:

[source,java]
----
public record EpisodeScheduledEvent(
    EpisodeId episodeId,
    EpisodeTitle title,
    LocalDate airDate
) implements DomainEvent {
}
----

Events are:

- Statements of fact (they have occurred)
- Immutable
- Can be replayed
- Enable event-driven architecture

Unlike **commands** (which can be rejected), events represent things that have already happened.

> == Hexagonal Architecture: Ports and Adapters

Reviewer: I'd get rid of `: Ports and Adapters`

Author: dropped.

public class EpisodeRepository implements PanacheRepository<EpisodeEntity> {
    
    public EpisodeAggregate persist(EpisodeAggregate aggregate) {
        EpisodeEntity entity = toEntity(aggregate);
        persist(entity);
        return rehydrate(entity);
    }
    
    public static EpisodeAggregate rehydrate(EpisodeEntity entity) {
        return new EpisodeAggregate(
            entity.id,
            new EpisodeTitle(entity.title),
            entity.description,
            new EpisodeAirDate(entity.airDate)
        );
    }
}
----

== The Trade-offs

=== More Code, Better Organization

Jeremy was upfront about the main drawback: **you'll write significantly more code** with DDD compared to a simple CRUD application. The demo showed:

- Value objects for each domain concept
- Separate DTOs for REST endpoints
- Commands for operations
- Events for notifications
- Mappers between layers

=== The Benefits

However, this additional code provides substantial benefits:

**1. Maintainability**: When inheriting code, business logic is easy to find - it's all in the aggregates.

**2. Testability**: Aggregates are POJOs that can be tested with plain JUnit, no framework required.

> **3. Flexibility**: Swapping persistence frameworks (e.g., from https://hibernate.org/orm/[Hibernate] to https://firebase.google.com/[Firebase]) only requires changing the repository implementation.

Reviewer 2: ```suggestion
**3. Flexibility**: Swapping persistence mechanisms (e.g., from a https://hibernate.org/orm/[Hibernate] relational database to a https://firebase.google.com/[Firebase] document store) only requires changing the repository implementation.
```

Firebase is a whole app framework, not an analogue to Hibernate, so clarifying a bit.

=== Transaction Boundaries

**Important**: Transaction boundaries should stay within each domain. If operations span multiple domains, consider:

- Using eventual consistency with domain events
- Implementing saga patterns for distributed transactions
- Carefully evaluating if you've drawn domain boundaries correctly

== Validation at Multiple Layers

Jeremy emphasized that validation occurs at different layers for different purposes:

**REST Layer (DTOs)**: Basic validation (not null, not blank)
[source,java]
----
public record CreateEpisodeRequest(
    @NotBlank String title,
    @NotBlank String description,
    @NotNull LocalDate airDate
) {}
----

**Domain Layer (Value Objects)**: Business rule validation
[source,java]
----
public record EpisodeAirDate(LocalDate value) {
    public EpisodeAirDate {
        if (value.isBefore(LocalDate.now())) {
            throw new IllegalArgumentException("Episode cannot air in the past");
        }
    }
}
----

**Application Layer**: Cross-aggregate validation
[source,java]
----
if (repository.titleExists(title)) {
    throw new IllegalArgumentException("Title already exists");
}
----

== AI-Assisted DDD Development

> Jeremy demonstrated using https://claude.ai/[Claude AI] to generate DDD boilerplate:

Reviewer 2: ```suggestion
Jeremy demonstrated using https://claude.ai/[Claude AI] to generate DDD boilerplate.
```

That colon seems odd before headings.

=== Skills and Specifications

He created:

- **Quarkus skills**: Instructions for using https://quarkus.io/guides/hibernate-orm-panache[Panache], REST endpoints, logging
- **DDD skills**: Patterns for aggregates, value objects, repositories
- **Specification files**: High-level requirements for the application

=== Results

Using AI with these skills, Jeremy generated a complete DDD application structure in "YOLO mode" (letting the AI run freely). While not perfect (it didn't use Panache as instructed), it created:

- Proper aggregate structure
- Value objects with validation
- Repository interfaces
- Domain events
- Application services

> The key insight: **DDD's structured, boilerplate-heavy nature makes it ideal for AI generation**, while keeping business logic centralized for human review.

Reviewer 2: ```suggestion
**DDD's structured, boilerplate-heavy nature makes it ideal for AI generation**, while keeping business logic centralized for human review.
```

=== Best Practices for AI-Assisted Development

Jeremy shared lessons learned:

1. **Set clear boundaries** in prompts (e.g., "always use GitOps, never imperative commands")

> 2. **Use structured prompting** like Architecture Definition Language (ADL) from https://www.developertoarchitect.com/[Mark Richards and Neal Ford]

Reviewer 2: Did he really say that? I totally don't remember it, but I'm very willing to believe I've missed it. 

But I do I wonder if it leaked forward from last week?

Reviewer 2: We may never know the truth without re-watching it, though. :)

Author: Before submitting this, I found other ghosts from last episode. I cross referenced this and the first two are ghosts as well. I couldn't find any reference searching the transcript from this episode to confirm the third lesson. I've opted to remove the section.

- Complex business logic
- Long-lived applications
- Multiple teams working on different domains
- Need for clear boundaries and maintainability
- AI-assisted development workflows

**Overkill**:
- Simple CRUD applications
- Prototypes or short-lived projects
- Small teams with simple requirements

== Resources and Next Steps

Jeremy offered to return for follow-up episodes covering:

- **State distribution** across domains
- **Event sourcing** and saga patterns
- **CQRS** for read models
- **AI-assisted DDD workflows** in depth

He also mentioned:

- Presenting at https://www.dddeurope.com/[Domain-Driven Design Europe]
- Speaking at https://exploreddd.com/[Explore DDD] in Colorado
- His blog at https://arrogantprogrammer.com/[The Arrogant Programmer]

== Key Takeaways for Developers

1. **DDD provides structure** that makes business logic easy to find and maintain
2. **Hexagonal architecture** keeps your domain pure and testable
3. **More code upfront** pays dividends in maintainability
4. **Ubiquitous language** bridges the gap between business and technology
5. **Value objects** enable validation at creation time
6. **Domain events** enable event-driven architecture naturally

> 7. **AI tools** work well with DDD's structured approach
> 8. **Transaction boundaries** should stay within domains
> 9. **Validation happens at multiple layers** for different purposes

Reviewer 2: ```suggestion
7. **Transaction boundaries** should stay within domains
8. **Validation happens at multiple layers** for different purposes
9. **AI tools** work well with DDD's structured approach
```

I like this order better :)
